internal/instance/resource_instance.go

```diff
@@ -942,22 +925,140 @@ func ToProfileListType(ctx context.Context, profiles []string) (types.List, diag
+		if st.Status == api.Running.String() && st.Processes <= 0 {
+			return st, "Running (initializing)", nil
+		}
+
 		return st, st.Status, nil
 	}
+
+	// Even though op.Wait has completed, wait until we can see
+	// the instance is fully started via a new API call.
+	stateConf := &retry.StateChangeConf{
+		Target:     []string{api.Running.String()},
+		Timeout:    3 * time.Minute,
+		MinTimeout: 3 * time.Second,
+		Delay:      refInterval,
+		Refresh:    instanceStartedCheck,
+	}
+
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.NewErrorDiagnostic(fmt.Sprintf("Failed to wait for instance %q to start", instanceName), err.Error())
+	}
+
+	return nil
 }
 
-// instanceNetworkStateRefreshFunc returns function that checks
-// whether instance has received an IP address.
-func instanceNetworkStateRefreshFunc(server lxd.InstanceServer, instanceName string) retry.StateRefreshFunc {
-	return func() (any, string, error) {
+// stopInstance stops an instance with the given name. It waits wither for
+// its status to become Stopped or for the instance to be not found (in case
+// of ephemeral instances).
+func stopInstance(ctx context.Context, server lxd.InstanceServer, instanceName string, refInterval time.Duration, force bool, ignoreNotFound bool) diag.Diagnostic {
+	is, etag, err := server.GetInstanceState(instanceName)
+	if err != nil {
+		return diag.NewErrorDiagnostic(fmt.Sprintf("Failed to retrieve state of instance %q", instanceName), err.Error())
+	}
+
+	// Return if already sopped.
+	if is.Status == api.Stopped.String() {
+		return nil
+	}
+
+	stopReq := api.InstanceStatePut{
+		Action:  "stop",
+		Force:   force,
+		Timeout: utils.ContextTimeout(ctx, 3*time.Minute),
+	}
+
+	// Stop the instance.
+	op, err := server.UpdateInstanceState(instanceName, stopReq, etag)
+	if err == nil {
+		err = op.WaitContext(ctx)
+	}
+
+	if err != nil {
+		return diag.NewErrorDiagnostic(fmt.Sprintf("Failed to stop instance %q", instanceName), err.Error())
+	}
+
+	instanceStoppedCheck := func() (any, string, error) {
+		st, _, err := server.GetInstanceState(instanceName)
+		if err != nil {
+			return st, "Error", err
+		}
+
+		return st, st.Status, nil
+	}
+
+	stateConf := &retry.StateChangeConf{
+		Target:     []string{api.Stopped.String()},
+		Timeout:    3 * time.Minute,
+		MinTimeout: 3 * time.Second,
+		Delay:      refInterval,
+		Refresh:    instanceStoppedCheck,
+	}
+
+	// Even though op.Wait has completed, wait until we can see
+	// the instance is stopped via a new API call.
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		if ignoreNotFound && errors.IsNotFoundError(err) {
+			return nil
+		}
+
+		return diag.NewErrorDiagnostic(fmt.Sprintf("Failed to wait for instance %q to stop", instanceName), err.Error())
+	}
+
+	return nil
+}
+
+// waitInstanceNetwork wait for an instance with the given name to receive
+// an IP address. This should be called only if we know the instance is
+// running.
```

Comment:
```suggestion
// waitInstanceNetwork waits for an instance with the given name to get
// an IPv4 address on any interface (!= `lo`). This should be called only if
// the instance's state is running.
```

`receive` makes sense in the the DHCP case, but since the IP can be statically configured, I feel that `get` might be slightly better even if more vague.